### PR TITLE
Decompile_touchscreen

### PR DESCRIPTION
**Pull request by PikalaxALT**
Retsam: src/gflib/touchpanel.c

### CODE REVIEW

File: src/berry_pots_app.c [hunk 1]
```diff
@@ -141,7 +141,7 @@ extern const UnkStruct_0200D2B4 ov17_02203EF8[2];
 extern const UnkStruct_0200D2B4 ov17_02203F48;
 extern const UnkStruct_0200D2B4 ov17_02203F70;
 extern const UnkStruct_0200D2B4 ov17_02203F98;
-extern const u8 ov17_02203FC0[4];
+extern const TouchscreenHitbox ov17_02203FC0;
```
Comment (adrienntindall) [inline src/berry_pots_app.c:144]: A lot of these touchscreen hitbox struct renames can likely be documented now if you wanted to go through quickly and see if there's anything obvious

File: src/touchscreen.c [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+                    return i;
+                }
+            }
+        }
+    }
+
+    return -1;
+}
+
+BOOL TouchscreenHitbox_TouchHeldIsIn(const TouchscreenHitbox *hitbox) {
```
Comment (adrienntindall) [inline src/touchscreen.c:83]: IsTouchHeldIn sounds better IMO, apply to other 'IsXIn(Y)' functions

Comment (adrienntindall) [inline src/touchscreen.c:83]: IsTouchHeldIn sounds better IMO, apply to other 'IsXIn(Y)' functions

File: src/berry_pots_app.c (hunk 1 above)
Comment (PikalaxALT) [reply src/berry_pots_app.c:144]: Most of them haven't hit C yet so ruling this out-of-scope for now.

File: src/touchscreen.c (hunk 2 above)
Comment (PikalaxALT) [reply src/touchscreen.c:83]: The use case would be `if (TouchscreenHitbox_TouchHeldIsIn(hitbox)) {`. The way it is now melds better with regular English grammar.

File: src/touchscreen.c [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+#include "system.h"
+
+static int TouchscreenHitbox_FindRectAtPoint(const TouchscreenHitbox *hitbox, u32 x, u32 y);
+static BOOL TouchscreenHitbox_PointIsInCircle(const TouchscreenHitbox *hitbox, u32 x, u32 y);
+static BOOL TouchscreenHitbox_PointIsInRect(const TouchscreenHitbox *hitbox, u32 x, u32 y);
+
+static int TouchscreenHitbox_FindRectAtPoint(const TouchscreenHitbox *hitboxes, u32 x, u32 y) {
+    // In this function and TouchscreenHitbox_PointIsInRect, GameFreak uses the bitwise-and operator (&) instead of the binary-and operator (&&).
+    // Other bad code decisions effectively mask this particular choice, albeit with a performance impact.
+    for (int i = 0; hitboxes[i].rect.top != TOUCHSCREEN_RECTLIST_END; ++i) {
```
Comment (adrienntindall) [inline src/touchscreen.c:12]: Is ++i necessary here?

File: src/touchscreen.c [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+int TouchscreenHitbox_FindHitboxAtTouchNew(const TouchscreenHitbox *hitboxes) {
+    if (gSystem.touchNew) {
+        for (int i = 0; hitboxes[i].circle.sentinel != TOUCHSCREEN_RECTLIST_END; ++i) {
+            if (hitboxes[i].circle.sentinel == TOUCHSCREEN_CIRCLE_SENTINEL) {
+                if (TouchscreenHitbox_PointIsInCircle(&hitboxes[i], gSystem.touchX, gSystem.touchY)) {
+                    return i;
+                }
+            } else {
+                if (TouchscreenHitbox_PointIsInRect(&hitboxes[i], gSystem.touchX, gSystem.touchY)) {
```
Comment (adrienntindall) [inline src/touchscreen.c:73]: else if can be merged into one statement

Comment (adrienntindall) [reply src/touchscreen.c:73]: etc for where it looks like you copied retsam code for the rest of the file

**Review by adrienntindall — APPROVED** [review]
if you're not wanting to document the arrays then really only minor syntax things, gonna leave approved so you can merge without having to wait after you're done

File: src/touchscreen.c (hunk 4 above)
Comment (PikalaxALT) [reply src/touchscreen.c:73]: I think this layout communicates the intention more clearly.

File: src/berry_pots_app.c [hunk 5]
```diff
@@ -1658,28 +1657,28 @@ static BOOL ov17_02203C20(BerryPotsAppData *data, BOOL *a1) {
 
 static int ov17_02203C78(BerryPotsAppData *data) {
     // touch screen coordinates?
-    u8 unkArray[4];
+    TouchscreenHitbox unkArray;
```
Comment (adrienntindall) [inline src/berry_pots_app.c:1660]: document unkArray

File: src/voltorb_flip/voltorb_flip_input.c [hunk 6]
```diff
@@ -230,7 +230,7 @@ static int ov122_021E9020(Ov122_021E8CFC *a0) {
     }
 
     if (a0->memoOpen && a0->unkD != 25 && a0->unkD != 26) {
-        u32 var3 = sub_02025224(ov122_021E9A1C);
+        u32 var3 = TouchscreenHitbox_FindRectAtTouchNew(ov122_021E9A1C);
```
Comment (adrienntindall) [inline src/voltorb_flip/voltorb_flip_input.c:233]: document var3?

**Review by adrienntindall — CHANGES_REQUESTED** [review]

File: src/touchscreen.c [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+
+int TouchscreenHitbox_FindHitboxAtTouchHeld(const TouchscreenHitbox *hitboxes) {
+    if (gSystem.touchHeld) {
+        for (int i = 0; hitboxes[i].circle.sentinel != TOUCHSCREEN_RECTLIST_END; ++i) {
+            if (hitboxes[i].circle.sentinel == TOUCHSCREEN_CIRCLE_SENTINEL) {
+                if (TouchscreenHitbox_PointIsInCircle(&hitboxes[i], gSystem.touchX, gSystem.touchY)) {
+                    return i;
+                }
+            } else {
+                if (TouchscreenHitbox_PointIsInRect(&hitboxes[i], gSystem.touchX, gSystem.touchY)) {
```
Comment (red031000) [inline src/touchscreen.c:55]: should be else if

File: src/touchscreen.c (hunk 4 above)
Comment (red031000) [reply src/touchscreen.c:73]: gotta agree here, should be else if

File: src/touchscreen.c [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+int TouchscreenHitbox_FindHitboxAtPoint(const TouchscreenHitbox *hitboxes, u32 x, u32 y) {
+    for (int i = 0; hitboxes[i].circle.sentinel != TOUCHSCREEN_RECTLIST_END; ++i) {
+        if (hitboxes[i].circle.sentinel == TOUCHSCREEN_CIRCLE_SENTINEL) {
+            if (TouchscreenHitbox_PointIsInCircle(&hitboxes[i], x, y)) {
+                return i;
+            }
+        } else {
+            if (TouchscreenHitbox_PointIsInRect(&hitboxes[i], x, y)) {
```
Comment (red031000) [inline src/touchscreen.c:142]: else if

**Review by red031000 — CHANGES_REQUESTED** [review]

File: src/touchscreen.c (hunk 8 above)
Comment (PikalaxALT) [reply src/touchscreen.c:142]: Strong disagree. Making this `else if` hides the similarity between these two blocks and hinders readability.

Comment (red031000) [reply src/touchscreen.c:142]: precisely because they are similar is why I think it should be else if, one is circle, one is rect

Comment (PikalaxALT) [reply src/touchscreen.c:142]: I'm sorry but your reasoning is self-contradictory.

Comment (red031000) [reply src/touchscreen.c:142]: I mean they seem like two cases of the same check, therefore they should be grouped together

File: src/touchscreen.c (hunk 7 above)
Comment (red031000) [reply src/touchscreen.c:55]: ok following discussion and careful re-reading, this is much better this way

**Review by red031000 — DISMISSED** [review]

**Review by red031000 — APPROVED** [review]

**Review by adrienntindall — APPROVED** [review]

**Pull request merged**